packages/bsky/src/auto-moderator/index.ts

```diff
@@ -0,0 +1,184 @@
+    if (abyssEndpoint && abyssPassword) {
+      this.takedownFlagger = new Abyss(abyssEndpoint, abyssPassword, ctx)
+    } else {
+      log.error(
+        { abyssEndpoint, abyssPassword },
+        'abyss not properly configured',
+      )
+    }
+
+    if (ctx.cfg.labelerPushUrl) {
+      const url = new URL(ctx.cfg.labelerPushUrl)
+      this.pushAgent = new AtpAgent({ service: url.origin })
+      this.pushAgent.api.setHeader(
+        'authorization',
+        buildBasicAuth(url.username, url.password),
+      )
+    }
+  }
+
+  processRecord(uri: AtUri, cid: CID, obj: unknown) {
+    this.ctx.backgroundQueue.add(async () => {
+      const { text, imgs } = getFieldsFromRecord(obj)
+      await Promise.all([
+        this.labelRecord(uri, cid, text, imgs).catch((err) => {
+          log.error(
+            { err, uri: uri.toString(), record: obj },
+            'failed to label record',
+          )
+        }),
+        this.checkImgForTakedown(uri, cid, imgs).catch((err) => {
+          log.error(
+            { err, uri: uri.toString(), record: obj },
+            'failed to check img for takedown',
+          )
+        }),
+      ])
+    })
+  }
+
+  async labelRecord(uri: AtUri, recordCid: CID, text: string[], imgs: CID[]) {
+    if (uri.collection === ids.AppBskyActorProfile) {
+      // @TODO label profiles
+      return
+    }
+    const allLabels = await Promise.all([
+      this.textLabeler?.labelText(text.join(' ')),
+      ...imgs.map((cid) => this.imgLabeler?.labelImg(uri.host, cid)),
+    ])
+    const labels = dedupe(allLabels.flat())
+    await this.storeLabels(uri, recordCid, labels)
+  }
+
+  async checkImgForTakedown(uri: AtUri, recordCid: CID, imgCids: CID[]) {
+    if (imgCids.length < 0) return
+    const results = await Promise.all(
+      imgCids.map((cid) => this.takedownFlagger?.scanImage(uri.host, cid)),
+    )
+    const takedownCids: CID[] = []
+    for (let i = 0; i < results.length; i++) {
+      if (results.at(i)?.length) {
+        takedownCids.push(imgCids[i])
+      }
+    }
+    if (takedownCids.length === 0) return
+    await this.ctx.db.transaction(async (dbTxn) => {
+      if (!this.services.moderation) {
+        log.error(
+          {
+            uri: uri.toString(),
+            cid: recordCid.toString(),
+            imgaes: takedownCids.map((c) => c.toString()),
+          },
+          'could not takedown flagged post',
+        )
+        return
+      }
+      const modSrvc = this.services.moderation(dbTxn)
+      const action = await modSrvc.logAction({
+        action: 'com.atproto.admin.defs#takedown',
+        subject: { uri, cid: recordCid },
+        subjectBlobCids: takedownCids,
+        reason: `automated takedown for labels: ${results.flat().join(', ')}`,
+        createdBy: this.ctx.cfg.labelerDid,
+      })
+      await modSrvc.takedownRecord({
+        takedownId: action.id,
+        uri: uri,
+        blobCids: takedownCids,
+      })
```

Comment:
For this to be synced with the PDS we want to actually hit the PDS and let it propagate to the appview—a similar approach can be found in `packages/bsky/src/db/periodic-moderation-action-reversal.ts`.

Reply:
ah good looks 👍